### PR TITLE
Fix: AfterGroups config annotation does not consider retries for tests 

### LINKED ISSUE #2684: AfterGroups config annotation does not consider retries for tests
`7.4.0`

### Expected behaviour

`@AfterGroups` configuration method should wait for all tests (including the retried tests) to finish.

### Actual behaviour
`@AfterGroups` configuration method runs before failed test retries are triggered.

### Is the issue reproducible on runner?

- [ ] Shell
- [X] Maven
- [ ] Gradle
- [ ] Ant
- [ ] Eclipse
- [X] IntelliJ
- [ ] NetBeans

```java
import static org.testng.Assert.assertTrue;

import java.util.ArrayList;
import java.util.List;
import org.testng.annotations.AfterGroups;
import org.testng.annotations.BeforeGroups;
import org.testng.annotations.Test;

public class SampleTestClassWithGroupConfigs {

  public static final List<String> logs = new ArrayList<>();

  @BeforeGroups("p1")
  public void beforeGroups() {
    logs.add("beforeGroups");
  }

  @Test(groups = "p1", retryAnalyzer = RerunAnalyzer.class)
  public void testMethod() {
    logs.add("testMethod");
    RerunAnalyzer.secondTestRetryCount++;
    assertTrue(RerunAnalyzer.secondTestRetryCount > RerunAnalyzer.maxRetryCount);
  }

  @AfterGroups("p1")
  public void afterGroups() {
    logs.add("afterGroups");
  }
}
```

```java
import org.testng.IRetryAnalyzer;
import org.testng.ITestResult;

public class RerunAnalyzer implements IRetryAnalyzer {

    public static final int maxRetryCount = 1;
    public static int secondTestRetryCount = 0;
    private int retryCount = 0;

    @Override
    public boolean retry(ITestResult iTestResult) {
        if (retryCount < maxRetryCount) {
            retryCount++;
            return true;
        }
        return false;
    }
}
```
```java
import static org.testng.Assert.assertEquals;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import org.testng.TestNG;
import org.testng.annotations.Test;
import org.testng.xml.XmlClass;
import org.testng.xml.XmlGroups;
import org.testng.xml.XmlRun;
import org.testng.xml.XmlSuite;
import org.testng.xml.XmlTest;

public class IssueTest {

  @Test
  public void testRetriedTestsWithGroupConfigs() {
    XmlSuite xmlSuite = new XmlSuite();
    xmlSuite.setName("groups_suite");
    xmlSuite.setVerbose(2);
    XmlTest xmlTest = new XmlTest(xmlSuite);
    xmlTest.setName("groups_test");
    XmlGroups xmlGroups = new XmlGroups();
    XmlRun xmlRun = new XmlRun();
    xmlRun.onInclude("p1");
    xmlGroups.setRun(xmlRun);
    xmlTest.setGroups(xmlGroups);
    xmlTest.setClasses(
        Collections.singletonList(
            new XmlClass(SampleTestClassWithGroupConfigs.class)
        )
    );
    TestNG testng = new TestNG();
    testng.setXmlSuites(Collections.singletonList(xmlSuite));
    testng.run();
    List<String> actual = SampleTestClassWithGroupConfigs.logs;
    List<String> expected = Arrays.asList("beforeGroups", "testMethod", "testMethod",
        "afterGroups");
    assertEquals(actual, expected, actual + " didn't match elements in expected list" + expected);
  }
}
```

### Contribution guidelines

Incase you plan to raise a pull request to fix this issue, please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

Reported from the [TestNG dev forum](https://groups.google.com/g/testng-dev/c/aPNlwrpWzjk/m/ymfbT6JaAwAJ)

### PR DESCRIPTION
**Pull request by morskyrv**
It fixes #2684 but I'm not sure about this solution, so waiting for comments.

Fixes #2684.

### Did you remember to?

- [x] Add test case(s)
- [x] Update `CHANGES.txt`
- [x] Auto applied styling via `./gradlew autostyleApply`

We encourage pull requests that:

* Add new features to TestNG (or)
* Fix bugs in TestNG

If your pull request involves fixing SonarQube issues then we would suggest that you please discuss this with the 
[TestNG-dev](https://groups.google.com/forum/#!forum/testng-dev) before you spend time working on it.

**Note:** For more information on contribution guidelines  please make sure you refer our [Contributing](.github/CONTRIBUTING.md) section for detailed set of steps.

### CODE REVIEW

File: testng-core/src/test/java/test/retryAnalyzer/issue2684/IssueTest.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import org.testng.xml.XmlClass;
+import org.testng.xml.XmlGroups;
+import org.testng.xml.XmlRun;
+import org.testng.xml.XmlSuite;
+import org.testng.xml.XmlTest;
+
+public class IssueTest {
+
+  @Test
+  public void testRetriedTestsWithGroupConfigs() {
```
Comment (krmahadevan) [inline testng-core/src/test/java/test/retryAnalyzer/issue2684/IssueTest.java:19]: Can you please move this test into `test.retryAnalyzer.RetryAnalyzerTest` and add a `description` attribute to the `@Test` which refers to `GITHUB-2684`

**Review by krmahadevan — CHANGES_REQUESTED** [review]
Also can you please help check if the problem is just confined to `@AfterGroups` or if this problem exists with other `@AfterXXX` configuration methods as well ?

File: testng-core/src/test/java/test/retryAnalyzer/issue2684/IssueTest.java (hunk 1 above)
Comment (morskyrv) [reply testng-core/src/test/java/test/retryAnalyzer/issue2684/IssueTest.java]: Done

**Comment by morskyrv** [comment]
@krmahadevan This bug affects only @AfterGroups, should I add checks for other config methods to the test too?
**Before Fix:**
image
**After Fix:**
image

**Comment by krmahadevan** [comment]
Yes pls. That way we will be sure that we have covered all aspects in terms of tests for retry analyser + config method combos.

**Comment by morskyrv** [comment]
@krmahadevan done

**Review by juherr — APPROVED** [review]

**Comment by krmahadevan** [comment]
@morskyrv Thank you for taking the time to help raise a PR. Hope to see many more from you. 👍

**Pull request merged**